source/vir/src/recursive_types.rs

```diff
@@ -587,6 +593,39 @@ pub fn check_traits(krate: &Krate, ctx: &GlobalCtx) -> Result<(), VirErr> {
     // This also ensures that whenever A is used in f and g,
     // the dictionary a: Dictionary_U<A> is available.
 
+    // To handle bounds on Self like this:
+    //   trait T: U {
+    //     fn f(x: Self, y: Self) -> bool;
+    //     fn g(x: Self, y: Self) -> Self { requires(f(x, y)); };
+    //   }
+    // We can store a Dictionary_U inside Dictionary_T:
+    //   struct Dictionary_T<Self> {
+    //     a: Dictionary_U<Self>,
+    //     f: Fn(x: Self, y: Self) -> bool,
+    //     g: Fn(x: Self, y: Self) -> Self { requires(f(x, y)); },
+    //   }
+    // This adds an edge, like for bounds on trait parameters:
+    //   - T --> U
+    // This also ensures that whenever Self is used in f and g,
+    // the dictionary a: Dictionary_U<Self> is available.
+
+    // To handle bounds on associated types:
+    //   trait T {
+    //     type Y: U;
+    //     fn f(x: Self, y: Y) -> bool;
+    //     fn g(x: Self, y: Y) -> Self { requires(f(x, y)); };
+    //   }
+    // We can store a Dictionary_U inside Dictionary_T:
+    //   struct Dictionary_T<Self, Y> {
+    //     a: Dictionary_U<Self::Y>,
```

Comment:
This is kind of my fault, because I forgot to comment the encoding of associated types, but I think `Y` should be a field of `Dictionary_T` rather than a field of `Self` or a type parameter to `Dictionary_T`.  This way, each trait gets to add its own associated types without disturbing `Self`, and users of `Dictionary_T` can read `Y` from the dictionary instead of having to figure out what `Y` is supposed to be as a type argument to the dictionary.